data/metric/metrics.go

```diff
@@ -14,112 +15,212 @@ import (
 // each one.
 type SourceType int
 
+// Attribute represents an attribute metric in key-value pair format.
+type Attribute struct {
+	Key   string
+	Value string
+}
+
 const (
 	// GAUGE is a value that may increase and decrease. It is stored as-is.
 	GAUGE SourceType = iota
-	// RATE is an ever-growing value which might be reseted. The package calculates the change rate.
+	// RATE is an ever-growing value which might be reset. The package calculates the change rate.
 	RATE SourceType = iota
-	// DELTA is an ever-growing value which might be reseted. The package calculates the difference between samples.
+	// DELTA is an ever-growing value which might be reset. The package calculates the difference between samples.
 	DELTA SourceType = iota
 	// ATTRIBUTE is any string value
 	ATTRIBUTE SourceType = iota
 )
 
+const (
+	// nsSeparator is the metric namespace separator
+	nsSeparator = "::"
+	// nsAttributeSeparator is the metric attribute key-value separator applied to generate the metric ns.
+	nsAttributeSeparator = "=="
+)
+
+// Errors
+var (
+	ErrNonNumeric        = errors.New("non-numeric value for rate/delta")
+	ErrNoStoreToCalcDiff = errors.New("cannot use deltas nor rates without persistent store")
+	ErrTooCloseSamples   = errors.New("samples too close in time, skipping")
+	ErrNegativeDiff      = errors.New("source was reset, skipping")
+	ErrOverrideSetAttrs  = errors.New("cannot overwrite metric-set attributes")
+	ErrDeltaWithNoAttrs  = errors.New("delta/rate metrics should be attached to an attribute identified metric-set")
+)
+
 // Set is the basic structure for storing metrics.
 type Set struct {
-	storer  persist.Storer
-	Metrics map[string]interface{}
+	storer       persist.Storer
+	Metrics      map[string]interface{}
+	nsAttributes []Attribute
 }
 
-// NewSet creates new metrics set.
-func NewSet(eventType string, storer persist.Storer) (*Set, error) {
-	ms := Set{
-		Metrics: map[string]interface{}{},
-		storer:  storer,
+// NewSet creates new metrics set, optionally related to a list of attributes. These attributes makes the metric-set unique.
+// If related attributes are used, then new attributes are added.
+// TODO remove obsolete returned error
```

Comment:
This is a BC from previous beta, but really nice to have as it removes a BC from SDKv2 =)